src/org/exist/xquery/value/ValueSequence.java

```diff
@@ -793,7 +793,7 @@ public Item nextItem() {
 
     private class ReverseValueSequenceIterator implements SequenceIterator {
 
-        private int pos = size - 1;
+        private int pos = size; // size is not the actual size
```

Comment:
The intention of the the fix is to solve something different. You have a point here, but I guess it is a bigger implementation/design issue? How can we solve this? (for an other PR).

for `add` and 'addAll' there are no problems, since it extends the internal array only, the Reverse-class is working from end to start.

For  `clear` you are right for sure.....